nifi-nar-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/cf/JndiJmsConnectionFactoryProvider.java

```diff
@@ -0,0 +1,165 @@
+        .name("java.naming.factory.initial")
+        .displayName("Initial Naming Factory Class")
+        .description("The fully qualified class name of the Java Initial Naming Factory (java.naming.factory.initial).")
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .required(true)
+        .build();
+    static final PropertyDescriptor NAMING_PROVIDER_URL = new Builder()
+        .name("java.naming.provider.url")
+        .displayName("Naming Provider URL")
+        .description("The URL of the JNDI Naming Provider to use")
+        .required(true)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+    static final PropertyDescriptor CONNECTION_FACTORY_NAME = new Builder()
+        .name("connection.factory.name")
+        .displayName("Connection Factory Name")
+        .description("The name of the JNDI Object to lookup for the Connection Factory")
+        .required(true)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .build();
+    static final PropertyDescriptor NAMING_FACTORY_LIBRARIES = new Builder()
+        .name("naming.factory.libraries")
+        .displayName("Naming Factory Libraries")
+        .description("Specifies .jar files or directories to add to the ClassPath in order to find the Initial Naming Factory Class")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .addValidator(StandardValidators.createListValidator(true, true, StandardValidators.createURLorFileValidator()))
+        .dynamicallyModifiesClasspath(true)
+        .build();
+
+    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Arrays.asList(
+        INITIAL_NAMING_FACTORY_CLASS,
+        NAMING_PROVIDER_URL,
+        CONNECTION_FACTORY_NAME,
+        NAMING_FACTORY_LIBRARIES);
+
+    private ConnectionFactory connectionFactory;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTY_DESCRIPTORS;
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new Builder()
+            .name(propertyDescriptorName)
+            .displayName(propertyDescriptorName)
+            .description("JNDI Initial Context Environment configuration for '" + propertyDescriptorName + "'")
+            .required(false)
+            .dynamic(true)
+            .addValidator(Validator.VALID)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
+    }
+
+    @OnDisabled
+    public void shutdown() {
+        connectionFactory = null;
+    }
+
+    @Override
+    public synchronized ConnectionFactory getConnectionFactory() {
```

Comment:
since this method is called each onTrigger I could be better to have two steps instantiation (to avoid sync for each call):

    @Override
    public ConnectionFactory getConnectionFactory() {
        return connectionFactory != null ? connectionFactory : _getConnectionFactory();
    }

    private synchronized ConnectionFactory _getConnectionFactory() {
        if (connectionFactory == null) {
            connectionFactory = lookupConnectionFactory();
        } else {
            getLogger().debug("Connection Factory has already been obtained from JNDI. Will return cached instance.");
        }

        return connectionFactory;
    }

Reply:
No, the above code is not thread safe. All access to the member variable must be protected. Synchronization is not a concern here. Either the method returns immediately or blocks while creating the connection factory, which is the desired behavior.